### PR TITLE
Support lookups for 'your' URIs

### PR DESCRIPTION
**Pull request by djmattyg007**
This adds the ability to utilise the "spotify:your:tracks" and
"spotify:your:albums" URIs through library lookups.

### CODE REVIEW

**Comment by djmattyg007** [comment]
I believe this should resolve #297

**Comment by marnunez** [comment]
It does! Thanks!

**Comment by kosciej16** [comment]
@djmattyg007 Is there a chance to merge it? Possibility of playing liked songs would be awesome!

File: mopidy_spotify/lookup.py [hunk 1]
```diff
@@ -92,3 +95,30 @@ def _lookup_playlist(config, session, web_client, uri):
...
+                web_track, bitrate=config["bitrate"]
+            )
+            if track is not None:
+                yield track
+    elif variant == "albums":
+        for item in items:
+            # The extra level here is to also support "saved album objects".
+            web_album = item.get("album", item)
+            if not translator.valid_web_data(web_album, "album"):
+                continue
```
Comment (kingosticks) [inline mopidy_spotify/lookup.py:121]:
```suggestion
            album_ref = translator.web_to_album_ref(web_album)
            if album_ref is None:
                continue
```

File: mopidy_spotify/lookup.py [hunk 2]
```diff
@@ -92,3 +95,30 @@ def _lookup_playlist(config, session, web_client, uri):
...
+
+
+def _lookup_your(config, session, web_client, uri):
+    parts = uri.replace("spotify:your:", "").split(":")
+    if len(parts) != 1:
+        return
+    variant = parts[0]
+
+    items = browse._load_your_music(web_client, variant)
+    if variant == "tracks":
```
Comment (kingosticks) [inline mopidy_spotify/lookup.py:107]: One thing different between the two varients here is that tracks won't be loaded by libspotify whereas albums will. I wonder if the difference is observable

**Review by kingosticks — APPROVED** [review]

File: mopidy_spotify/lookup.py [hunk 3]
```diff
@@ -92,3 +95,30 @@ def _lookup_playlist(config, session, web_client, uri):
...
+            )
+            if track is not None:
+                yield track
+    elif variant == "albums":
+        for item in items:
+            # The extra level here is to also support "saved album objects".
+            web_album = item.get("album", item)
+            if not translator.valid_web_data(web_album, "album"):
+                continue
+            sp_link = session.get_link(web_album.get("uri"))
```
[the remark this answers is not in the document]
Comment (djmattyg007) [reply mopidy_spotify/lookup.py:122]: Would you be able to explain what the meaningful difference is between the two versions of this code?

File: mopidy_spotify/lookup.py (hunk 2 above)
Comment (djmattyg007) [reply mopidy_spotify/lookup.py:107]: Would you be able to elaborate here? I'm pretty sure I tested this quite extensively at the time I wrote it. Unless something has changed with some other component, it should work.

Comment (kingosticks) [reply mopidy_spotify/lookup.py:107]: When you load any sort of container thing in libspotify it loads the track data for that in the background so if you then go and use the track (I.e. play it) it should happen instantly. If the tracks are being loaded via the Web API, when you go to play it that'll be the first time libspotify hears about the track and first has to load the data. There might be a short pause for that but I wasn't sure it'd be noticeable.

File: mopidy_spotify/lookup.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kingosticks) [reply mopidy_spotify/lookup.py:122]: Using the existing function hides the internal Web API structure and does some error checking for you. I appreciate you still need to know how the response is structured later on but this is one less place. I also have so far preferred not to called `valid_web_data` outside of the translator (although maybe we do this already, I can't remember and I've not checked)

File: mopidy_spotify/lookup.py (hunk 2 above)
Comment (djmattyg007) [reply mopidy_spotify/lookup.py:107]: Is there an alternative that would avoid that potential pause, and would it be worth scrapping what I've worked on in this PR to implement it vs merging this now and implementing that later?

Comment (kingosticks) [reply mopidy_spotify/lookup.py:107]: The only way I know (and it's a little hacky) is to do something like [this](https://github.com/mopidy/mopidy-spotify/blob/master/mopidy_spotify/playlists.py#L98-L110). I'll test it myself later but if you didn't notice any issue it's probably fine. I'd be happy to merge as it is and we can always come back later if people find the performance is an actual problem.

Comment (djmattyg007) [reply mopidy_spotify/lookup.py:107]: I did wonder what the purpose of that code was.

Either way, I definitely think this feature is better off existing in a non-perfect state than not existing at all. I'll make the translator-related change you suggested to the lookup code at some point over the next couple of days, and then we can get this merged :+1:

**Comment by djmattyg007** [comment]
I've updated this PR. It should be good to go :+1:

**Comment by kingosticks** [comment]
Cool, so as a resulf of this we can now do:
* `core.library.lookup(uris=["spotify:your:tracks"])`
* `core.library.search(query={"uris": ["spotify:your:tracks"]})`

e.g.
```
curl -d '{"jsonrpc": "2.0", "id": 1, "method": "core.library.lookup", "params": {"uris": ["spotify:your:tracks"]}}' -H 'Content-Type: application/json' http://localhost:6680/mopidy/rpc

curl -d '{"jsonrpc": "2.0", "id": 1, "method": "core.library.search", "params": {"query": {"uri": ["spotify:your:tracks"]}}}' -H 'Content-Type: application/json' http://localhost:6680/mopidy/rpc
```

Perhaps we can use this functionality when we fix https://github.com/mopidy/mopidy-spotify/issues/328 (in a later PR).

**Pull request merged**